rewriter/GenCallAsm.cpp

```diff
@@ -280,21 +280,309 @@ static std::string sig_string(const CAbiSignature &sig,
+  // Copy any stack returns to caller's stack
+  if (stack_return_size > 0) {
+    // If the return value is in memory, we pushed the initial rdi (which
+    // pointed to the caller's return value memory) onto the compartment's
+    // stack. Now we pop it into rax which is where the address of the return
+    // value must be on return.
+    add_asm_line(aw, "popq %rax");
+
+    // After the pop rsp points to the memory for the return value on the
+    // compartment's stack.
+    add_comment_line(aw,
+                     "Copy "s + std::to_string(stack_return_size) +
+                         " bytes for the return value to the caller's stack");
+    for (int i = 0; i < stack_return_size; i += 8) {
+      add_asm_line(aw, "popq "s + std::to_string(i) + "(%rax)");
+    }
+
+    if (stack_return_padding > 0) {
+      add_asm_line(aw, "addq $8, %rsp");
+    }
+  }
+
+  // Switch back to the caller's stack
+  emit_switch_stacks(aw, target_pkey, caller_pkey, "r11"s);
+}
+
+static void finalize_and_return_to_caller(AsmWriter &aw, uint32_t target_pkey, uint32_t caller_pkey, const std::vector<ParamLocation> &return_locs, Arch arch) {
+  if (arch == Arch::X86) {
+    // After calling the target function, we only need to zero out registers not
+    // used for return values if the target is a protected compartment
+    if (target_pkey != 0) {
+      add_comment_line(
+          aw, "Push return regs to caller's stack before scrubbing registers");
+      // Push return regs to the caller's stack. These pushes have matching pops
+      // before the return so it has no net effect on the caller's stack pointer.
+      for (auto loc : return_locs) {
+        if (!loc.is_stack()) {
+          emit_reg_push(aw, loc);
+        }
+      }
+
+      // Zero all registers except rsp
+      add_comment_line(aw, "Scrub registers after call");
+      // FIXME: If this will use the System V ABI make sure that the %rsp is aligned
+      // before this call
+      add_asm_line(aw, "call __libia2_scrub_registers");
+
+      // pop return regs
+      add_comment_line(aw, "Pop return regs");
+      for (auto loc = return_locs.rbegin(); loc != return_locs.rend(); loc++) {
+        if (!loc->is_stack()) {
+          emit_reg_pop(aw, *loc);
+        }
+      }
+    }
+    // Once again use r10 and r11 as scratch registers
+    add_comment_line(aw, "Set PKRU to the caller's value");
+    add_asm_line(aw, "movq %rax, %r10");
+    add_asm_line(aw, "movq %rdx, %r11");
+    emit_wrpkru(aw, caller_pkey);
+    add_asm_line(aw, "movq %r10, %rax");
+    add_asm_line(aw, "movq %r11, %rdx");
+
+    // Load registers that are preserved across function calls after switching
+    // back to the caller's compartment's stack. This is on the caller's stack so
+    // it's not in the diagram above.
+    for (auto r = preserved_registers.rbegin(); r != preserved_registers.rend();
+         r++) {
+      add_asm_line(aw, "popq %"s + *r);
+    }
+    // Restore the caller's frame pointer
```

Comment:
The comment about the function prologue also applies to the epilogue here.

Reply:
Moved this into a corresponding `emit_restore_regs`